### PR TITLE
Openemr fhir medication request

### PR DESCRIPTION
**Pull request by adunsulag**
MedicationRequest FHIR implementation

TL;DR Changed edit issues screen to accommodate additional medication fields.  Implemented FHIR MedicationRequest. 

Note: For ONC Testing we need to update our test database to include at least 4 medication issues that support the following combos (intent=order,status=active|intent=order,status=completed|intent=plan,status=active|intent=plan,status=completed). We also need to have at least one drug_dosage_instructions filled out.

This was a much bigger endeavor than expected due to the way OpenEMR handles medications and the way FHIR handles medications.  
US Core combines the listing of a patient's medications and the patient's perscriptions into a single MedicationRequest resource.  FHIR R4 has a breakdown of various resources but US Core brings them into a single resource.  Many of the US core must support fields are not supported in the NewCrop and WENO e-prescribe and so we add those fields to the Medications issues screen.

In this commit I add to the Medication Issues screen the ability to specify a medication order's intent, the usage category of a medication, and an open text field for dosage instructions for a patient taking the medication.  

Screenshots of that are here:
image

Request Intent List options
image

Category Usage List options
image

The usage category specifies where the medication is intended to be used (at home, in an inpatient, outpatient, or at the time of discharge).  Eventually its been proposed to move this into the procedure order screen which will have to be a project by someone else. 

I've gone and created a lists_medication table to hold all of the medication specific fields for an issue there.  I ended up moving the patient issue saving to its own service so we could centralize this logic and make sure we are saving data consistently.  I had to modify the deleter.php file to make sure we delete the data properly.  All the extra medication data is optional in the database so we don't break any existing logic elsewhere where we utilize issues.  I noticed the add_edit_issue.php was using our old system of escaping columns and I moved the logic to use parameterized sql binding.

I had to fix some bugs in the BaseService.php classes dealing with NULL values.

More green checkmarks!
image

### CODE REVIEW

**Comment by adunsulag** [comment]
So my psr-12 checker doesn't find any styling issues... 
image

Seems like my lot in life is to always fail something on a pull request lol.

**Comment by sjpadgett** [comment]
I used to have an issue with Travis however, I went out of my way to make friends with git actions. Ya might want to consider reaching out and making a friend of git!:)

**Comment by stephenwaite** [comment]
hi @adunsulag , it's `openemr-cmd pr` for the src directory, think we need special handling for some of the older classes

**Comment by adunsulag** [comment]
@stephenwaite Thank you that was very helpful.  I didn't realize the pr and pf operated off of different rules.  I was able to duplicate the style issues found in the pr report and hopefully the style checker now passes :)

File: interface/patient_file/summary/add_edit_issue.php [hunk 1]
```diff
@@ -230,86 +231,49 @@ function ActiveIssueCodeRecycleFn($thispid2, $ISSUE_TYPES2)
...
-            sqlStatement(
-                'UPDATE prescriptions SET '
-                . 'medication = 0 where patient_id = ? '
-                . " and upper(trim(drug)) = ? "
-                . ' and medication = 1',
-                array($thispid, strtoupper($_POST['form_title']))
-            );
+    $issueRecord = [
+        'type' => $text_type
+        // I don't like how these date columns use this quasi 'NULL' value but its how the underlying service works...
```
Comment (bradymiller) [inline interface/patient_file/summary/add_edit_issue.php:243]: The original sql queries are using null without quotes though (they are run the the QuotedOrNull() function to ensure that the null does not get the quotes).

Comment (bradymiller) [reply interface/patient_file/summary/add_edit_issue.php:243]: oh, that QuotedOrNull() function is weird.
It doesn't seem right to do that. I wonder if that got wrongly updated during sql-injection work in past.

Comment (bradymiller) [reply interface/patient_file/summary/add_edit_issue.php:243]: nope, I was wrong. That return "NULL" code has been there since the beginning of time.

File: sql/6_0_0-to-6_1_0_upgrade.sql [hunk 2]
```diff
@@ -850,3 +850,51 @@ ALTER TABLE `form_vitals` ADD `inhaled_oxygen_concentration` float(4,1) DEFAULT
 #EndIf
 
 UPDATE `list_options` SET `notes` = 'LOINC:11502-2' WHERE `list_options`.`list_id` = 'Clinical_Note_Type' AND `list_options`.`option_id` = 'laboratory_report_narrative';
+
+
+#IfMissingColumn prescriptions drug_dosage_instructions
+ALTER TABLE `prescriptions` ADD COLUMN drug_dosage_instructions longtext NULL DEFAULT NULL COMMENT 'Medication dosage instructions';
```
Comment (bradymiller) [inline sql/6_0_0-to-6_1_0_upgrade.sql:856]: make text fields bare (ok for the COMMENT, though); just trying to keep the text fields consistent since have spent resources in past standardizing these.

File: sql/6_0_0-to-6_1_0_upgrade.sql [hunk 3]
```diff
@@ -850,3 +850,51 @@ ALTER TABLE `form_vitals` ADD `inhaled_oxygen_concentration` float(4,1) DEFAULT
...
+#EndIf
+
+#IfMissingColumn prescriptions usage_category
+ALTER TABLE `prescriptions` ADD COLUMN `usage_category` VARCHAR(100) NULL COMMENT 'option_id in list_options.list_id=medication-usage-category';
+ALTER TABLE `prescriptions` ADD COLUMN `usage_category_title` VARCHAR(255) NOT NULL COMMENT 'title in list_options.list_id=medication-usage-category';
+ALTER TABLE `prescriptions` ADD COLUMN `request_intent` VARCHAR(100) NULL COMMENT 'option_id in list_options.list_id=medication-request-intent';
+ALTER TABLE `prescriptions` ADD COLUMN `request_intent_title` VARCHAR(255) NOT NULL COMMENT 'title in list_options.list_id=medication-request-intent';
```
Comment (bradymiller) [inline sql/6_0_0-to-6_1_0_upgrade.sql:863]: Do the titles need to be stored? (or can just collect them from list_options when needed)

File: sql/6_0_0-to-6_1_0_upgrade.sql [hunk 4]
```diff
@@ -850,3 +850,51 @@ ALTER TABLE `form_vitals` ADD `inhaled_oxygen_concentration` float(4,1) DEFAULT
...
+#EndIf
+
+#IfNotTable lists_medication
+CREATE TABLE `lists_medication` (
+    `id` BIGINT(20) NOT NULL AUTO_INCREMENT
+    , `list_id` BIGINT(20) NULL COMMENT 'FK Reference to lists.id'
+    , `drug_dosage_instructions` LONGTEXT NULL COMMENT 'Free text dosage instructions for taking the drug'
```
Comment (bradymiller) [inline sql/6_0_0-to-6_1_0_upgrade.sql:870]: make text fields bare (ok for the COMMENT, though); just trying to keep the text fields consistent since have spent resources in past standardizing these.

File: sql/6_0_0-to-6_1_0_upgrade.sql [hunk 5]
```diff
@@ -850,3 +850,51 @@ ALTER TABLE `form_vitals` ADD `inhaled_oxygen_concentration` float(4,1) DEFAULT
...
+INSERT INTO list_options (list_id,option_id,title, seq, is_default, option_value, notes) VALUES ('lists','medication-usage-category','Medication Usage Category',0, 1, 0, 'Values taken from http://hl7.org/fhir/R4/valueset-medicationrequest-category.html');
+INSERT INTO list_options (list_id,option_id,title,seq,is_default,activity, notes) VALUES ('medication-usage-category','inpatient','Inpatient',10,0,1, 'Includes requests for medications to be administered or consumed in an inpatient or acute care setting');
+INSERT INTO list_options (list_id,option_id,title,seq,is_default,activity, notes) VALUES ('medication-usage-category','outpatient','Outpatient',20,0,1, 'Includes requests for medications to be administered or consumed in an outpatient setting (for example, Emergency Department, Outpatient Clinic, Outpatient Surgery, Doctor''s office)');
+INSERT INTO list_options (list_id,option_id,title,seq,is_default,activity, notes) VALUES ('medication-usage-category','community','Home/Community',30,1,1, 'Includes requests for medications to be administered or consumed by the patient in their home (this would include long term care or nursing homes, hospices, etc.)');
```
Comment (bradymiller) [inline sql/6_0_0-to-6_1_0_upgrade.sql:886]: note that assigning this as default means it will be autoselected rather than the unassigned being autoselected (thus will basically always be autoselected in gui rather than actually being chosen by the user)

File: sql/6_0_0-to-6_1_0_upgrade.sql [hunk 6]
```diff
@@ -850,3 +850,51 @@ ALTER TABLE `form_vitals` ADD `inhaled_oxygen_concentration` float(4,1) DEFAULT
...
+#EndIf
+
+#IfNotRow2D list_options list_id lists option_id medication-request-intent
+INSERT INTO list_options (list_id,option_id,title, seq, is_default, option_value, notes) VALUES ('lists','medication-request-intent','Medication Request Intent',0, 1, 0, 'Values taken from http://hl7.org/fhir/R4/valueset-medicationrequest-intent.html');
+INSERT INTO list_options (list_id,option_id,title,seq,is_default,activity, notes) VALUES ('medication-request-intent','proposal','Proposal',10,0,1, 'The request is a suggestion made by someone/something that doesn''t have an intention to ensure it occurs and without providing an authorization to act ');
+INSERT INTO list_options (list_id,option_id,title,seq,is_default,activity, notes) VALUES ('medication-request-intent','plan','Plan',20,0,1, 'The request represents an intention to ensure something occurs without providing an authorization for others to act.');
+INSERT INTO list_options (list_id,option_id,title,seq,is_default,activity, notes) VALUES ('medication-request-intent','order','Order',30,1,1, 'The request represents a request/demand and authorization for action');
```
Comment (bradymiller) [inline sql/6_0_0-to-6_1_0_upgrade.sql:894]: note that assigning this as default means it will be autoselected rather than the unassigned being autoselected (thus will basically always be autoselected in gui rather than actually being chosen by the user)

File: sql/6_0_0-to-6_1_0_upgrade.sql [hunk 7]
```diff
@@ -850,3 +850,51 @@ ALTER TABLE `form_vitals` ADD `inhaled_oxygen_concentration` float(4,1) DEFAULT
...
+INSERT INTO list_options (list_id,option_id,title,seq,is_default,activity, notes) VALUES ('medication-usage-category','discharge','Discharge',40,0,1, 'Includes requests for medications created when the patient is being released from a facility');
+#EndIf
+
+#IfNotRow2D list_options list_id lists option_id medication-request-intent
+INSERT INTO list_options (list_id,option_id,title, seq, is_default, option_value, notes) VALUES ('lists','medication-request-intent','Medication Request Intent',0, 1, 0, 'Values taken from http://hl7.org/fhir/R4/valueset-medicationrequest-intent.html');
+INSERT INTO list_options (list_id,option_id,title,seq,is_default,activity, notes) VALUES ('medication-request-intent','proposal','Proposal',10,0,1, 'The request is a suggestion made by someone/something that doesn''t have an intention to ensure it occurs and without providing an authorization to act ');
```
Comment (bradymiller) [inline sql/6_0_0-to-6_1_0_upgrade.sql:892]: meant to end notes with whitespace?

File: sql/6_0_0-to-6_1_0_upgrade.sql [hunk 8]
```diff
@@ -850,3 +850,51 @@ ALTER TABLE `form_vitals` ADD `inhaled_oxygen_concentration` float(4,1) DEFAULT
...
+INSERT INTO list_options (list_id,option_id,title,seq,is_default,activity, notes) VALUES ('medication-request-intent','plan','Plan',20,0,1, 'The request represents an intention to ensure something occurs without providing an authorization for others to act.');
+INSERT INTO list_options (list_id,option_id,title,seq,is_default,activity, notes) VALUES ('medication-request-intent','order','Order',30,1,1, 'The request represents a request/demand and authorization for action');
+INSERT INTO list_options (list_id,option_id,title,seq,is_default,activity, notes) VALUES ('medication-request-intent','original-order','Original Order',40,0,1, 'The request represents the original authorization for the medication request.');
+INSERT INTO list_options (list_id,option_id,title,seq,is_default,activity, notes) VALUES ('medication-request-intent','reflex-order','Reflex Order',50,0,1, 'The request represents an automatically generated supplemental authorization for action based on a parent authorization together with initial results of the action taken against that parent authorization..');
```
Comment (bradymiller) [inline sql/6_0_0-to-6_1_0_upgrade.sql:896]: meant to end notes with 2 periods?

File: sql/database.sql [hunk 9]
```diff
@@ -6562,12 +6562,30 @@ CREATE TABLE `lists` (
   `severity_al` VARCHAR( 50 ) DEFAULT NULL,
   `external_id` VARCHAR(20) DEFAULT NULL,
   `list_option_id` VARCHAR(100) DEFAULT NULL COMMENT 'Reference to list_options table',
+  `drug_dosage_instructions` longtext DEFAULT NULL COMMENT 'Medication list dosage instructions',
```
Comment (bradymiller) [inline sql/database.sql:6565]: make text fields bare (ok for the COMMENT, though); just trying to keep the text fields consistent since have spent resources in past standardizing these.

File: sql/database.sql [hunk 10]
```diff
@@ -6562,12 +6562,30 @@ CREATE TABLE `lists` (
...
   UNIQUE KEY `uuid` (`uuid`)
 ) ENGINE=InnoDB AUTO_INCREMENT=1;
 
+
+-- --------------------------------------------------------
+DROP TABLE IF EXISTS `lists_medication`;
+CREATE TABLE `lists_medication` (
+                                    `id` BIGINT(20) NOT NULL AUTO_INCREMENT
+    , `list_id` BIGINT(20) NULL COMMENT 'FK Reference to lists.id'
+    , `drug_dosage_instructions` LONGTEXT NULL COMMENT 'Free text dosage instructions for taking the drug'
```
Comment (bradymiller) [inline sql/database.sql:6578]: make text fields bare (ok for the COMMENT, though); just trying to keep the text fields consistent since have spent resources in past standardizing these.

File: sql/database.sql [hunk 11]
```diff
@@ -10848,6 +10870,25 @@ VALUES
...
+INSERT INTO list_options (list_id,option_id,title,seq,is_default,activity, notes) VALUES ('medication-request-intent','filler-order','Filler Order',60,0,1, 'The request represents the view of an authorization instantiated by a fulfilling system representing the details of the fulfiller''s intention to act upon a submitted order.');
+INSERT INTO list_options (list_id,option_id,title,seq,is_default,activity, notes) VALUES ('medication-request-intent','instance-order','Instance Order',70,0,1, 'The request represents an instance for the particular order, for example a medication administration record.');
+INSERT INTO list_options (list_id,option_id,title,seq,is_default,activity, notes) VALUES ('medication-request-intent','option','Option',80,0,1, 'The request represents a component or option for a RequestGroup that establishes timing, conditionality and/or other constraints among a set of requests.');
```
Comment (bradymiller) [inline sql/database.sql:10890]: see the sql upgrade scripts for comments on above 2 lists

File: sql/6_0_0-to-6_1_0_upgrade.sql (hunk 3 above)
Comment (bradymiller) [reply sql/6_0_0-to-6_1_0_upgrade.sql:863]: disregard above post. you answered it in a code comment below :)

File: src/Services/PrescriptionService.php [hunk 12]
```diff
@@ -78,70 +80,195 @@ public function getAll($search = array(), $isAndCondition = true, $puuidBind = n
...
+                ) encounter
+                ON encounter.encounter = combined_prescriptions.encounter
+                LEFT JOIN (
+                    SELECT 
+                           id AS practitioner_id
+                           ,uuid AS pruuid
+                    FROM users
+                    WHERE users.npi IS NOT NULL AND users.npi != ''
+                ) practitioner
+                ON practitioner.practitioner_id = combined_prescriptions.provider_id";
```
Comment (bradymiller) [inline src/Services/PrescriptionService.php:240]: dang, now that is a sql query!

2DV-2

**Comment by bradymiller** [comment]
code review complete (i didn't test) :)

File: interface/patient_file/summary/add_edit_issue.php (hunk 1 above)
Comment (adunsulag) [reply interface/patient_file/summary/add_edit_issue.php:243]: Just want to make sure from your comment, this is ok to do it the way I'm doing here right?  The values still get set to NULL when they are saved.

File: sql/6_0_0-to-6_1_0_upgrade.sql (hunk 5 above)
Comment (adunsulag) [reply sql/6_0_0-to-6_1_0_upgrade.sql:886]: The category is required with no option to null-flavor it for FHIR so I need it to be chosen by default.  The default value should be Home/Community.

File: sql/6_0_0-to-6_1_0_upgrade.sql (hunk 6 above)
Comment (adunsulag) [reply sql/6_0_0-to-6_1_0_upgrade.sql:894]: The value is required to be filled out, Based on conversations with @juggernautsei and @kchapple I believe this should default to 'Order'.

File: sql/6_0_0-to-6_1_0_upgrade.sql (hunk 7 above)
Comment (adunsulag) [reply sql/6_0_0-to-6_1_0_upgrade.sql:892]: Nope it was copied from HL7's docs. Removed the whitespace.

File: sql/6_0_0-to-6_1_0_upgrade.sql (hunk 8 above)
Comment (adunsulag) [reply sql/6_0_0-to-6_1_0_upgrade.sql:896]: Nope it was copied from HL7's docs. Removed the extra period.

**Comment by adunsulag** [comment]
@bradymiller I believe I've covered everything you mentioned.  If it looks good feel free to bring it in.

**Comment by sjpadgett** [comment]
@adunsulag I restarted actions for ya. Also, I believe you may have git privilege's by now so bring in when ready. If not, i'll do it.

**Comment by adunsulag** [comment]
@sjpadgett I hadn't responded to Brady's email about getting git privileges until just a bit ago so if you are good with bringing it in that'd be great.  Wasn't sure if Brady wanted another pass to look at it again.

**Comment by stephenwaite** [comment]
i better take a look too :)

**Comment by adunsulag** [comment]
@stephenwaite That's fine, just let me know if you seen any problems.

**Comment by stephenwaite** [comment]
just kidding of course, great work!


https://user-images.githubusercontent.com/5618372/128774629-63b6155b-5e7e-4a61-9cc2-06f783b34fbc.mp4

**Comment by sjpadgett** [comment]
I don't care how diligent I am, ole eagle eye(Brady) tends to find i've missed something. Tho after several years, i'm getting better...

**Comment by sjpadgett** [comment]
I went through and looks like ya got his items so bringing in.
Again, thanks for the nice code Stephen. I love to watch you work.

**Comment by adunsulag** [comment]
@sjpadgett Thank you for doing that.  Glad to hear you appreciate the code :)

File: interface/patient_file/summary/add_edit_issue.php (hunk 1 above)
Comment (bradymiller) [reply interface/patient_file/summary/add_edit_issue.php:243]: yep, looks good
